src/shaka_scheme/runtime/stdproc/pairs_and_lists.cpp

```diff
@@ -0,0 +1,428 @@
+                                    "procedure");
+  }
+  return {create_node(Data(DataPair(Data(*args[0]),
+                               Data(*args[1]))))
+  };
+}
+
+//(car ...)
+Args car(Args args) {
+
+  if (args[0]->get_type() != Data::Type::DATA_PAIR) {
+    throw InvalidInputException(3,
+                                "car: Type of given argument is not "
+                                    "DATA_PAIR");
+  }
+  if (args[0]->get_type() == Data::Type::NULL_LIST) {
+    throw InvalidInputException(4,
+                                "car: Cannot return the car of"
+                                    "an empty list");
+  }
+  DataPair dp{create_node(Data(
+      DataPair(args[0])))};
+  return {create_node(Data(dp.car()))};
+}
+
+//(cdr ...)
+Args cdr(Args args) {
+
+  if (args[0]->get_type() != Data::Type::DATA_PAIR) {
+    throw InvalidInputException(5,
+                                "cdr: Type of given argument is not "
+                                    "DATA_PAIR");
+  }
+  if (args[0]->get_type() == Data::Type::NULL_LIST) {
+    throw InvalidInputException(6,
+                                "cdr: Cannot return the cdr of"
+                                    "an empty list");
+  }
+  DataPair dp{create_node(Data(
+      DataPair(args[0])))};
+  return {create_node(Data(dp.cdr()))};
+}
+
+//(set-car! ...)
+Args set_car(Args args) {
+  if (args[0]->get_type() != Data::Type::DATA_PAIR) {
+    throw InvalidInputException(10,
+                                "set-car!: Type of given argument is not "
+                                    "DATA_PAIR");
+  }
+  if (args.size() != 2) {
+    throw InvalidInputException(11,
+                                "set-car!: Invalid number of arguments given");
+  }
+  args[0]->get<DataPair>().set_car(args[1]);
+  return {create_unspecified()};
+}
+
+//(set-cdr! ...)
+Args set_cdr(Args args) {
+  if (args[0]->get_type() != Data::Type::DATA_PAIR) {
+    throw InvalidInputException(12,
+                                "set-cdr!: Type of given argument is not "
+                                    "DATA_PAIR");
+  }
+  if (args.size() != 2) {
+    throw InvalidInputException(13,
+                                "set-cdr!: Invalid number of arguments given");
+  }
+  args[0]->get<DataPair>().set_cdr(args[1]);
+  return {create_unspecified()};
+}
+
+//(null? ...)
+Args is_null(Args args) {
+  if (args.size() != 1) {
+    throw InvalidInputException(7,
+                                "null?: Invalid number of arguments for "
+                                    "procedure");
+  }
+  return {create_node(Data(Boolean(
+      args[0]->get_type() == Data::Type::NULL_LIST)
+  ))};
+}
+
+//(list? ...)
+Args is_list(Args args) {
+  if (args.size() != 1) {
+    throw InvalidInputException(8,
+                                "list?: Invalid number of arguments for "
+                                    "procedure");
+  }
+  return {create_node(Data(Boolean(

```

Comment:
This is wrong. The null list is also a list but it is not a pair.

Reply:
Changed